src/lib/coreaudio/CoreAudioDevice.cxx

```diff
@@ -0,0 +1,676 @@
+		FormatError(macos_output_domain, "Unable to remove IOProc. Error = %s", GetError(err).c_str());
+
+	io_proc = nullptr; // Clear the reference no matter what
+	
+	return true;
+}
+
+std::string
+CoreAudioDevice::GetName() const
+{
+	if (!device_id)
+		return "";
+
+	AudioObjectPropertyAddress  aopa;
+	aopa.mScope    = kAudioDevicePropertyScopeOutput;
+	aopa.mElement  = 0;
+	aopa.mSelector = kAudioDevicePropertyDeviceNameCFString;
+
+	std::string name;
+	CFStringRef deviceName = nullptr;
+	UInt32 propertySize = sizeof(deviceName);
+
+	OSStatus err = AudioObjectGetPropertyData(device_id, &aopa, 0, NULL, &propertySize, &deviceName);
+
+	if (err != noErr)
+	{
+		FormatError(macos_output_domain, "Unable to get device name - id: 0x%04x. Error = %s", (uint)device_id, GetError(err).c_str());
+	}
+	else
+	{
+		// The +1 is for having space for the string to be NUL terminated
+		CFIndex bufferSize = CFStringGetLength(deviceName) + 1;
+		char buffer[bufferSize];
+		
+		if (CFStringGetCString(deviceName, buffer, bufferSize, kCFStringEncodingUTF8))
+		{
+			name.assign(buffer);
+		}
+		CFRelease(deviceName);
+	}
+	return name;
+}
+
+UInt32
+CoreAudioDevice::GetTotalOutputChannels() const
+{
+	UInt32 channels = 0;
+
+	if (!device_id)
+		return channels;
+
+	AudioObjectPropertyAddress  aopa;
+	aopa.mScope    = kAudioDevicePropertyScopeOutput;
+	aopa.mElement  = 0;
+	aopa.mSelector = kAudioDevicePropertyStreamConfiguration;
+
+	UInt32 size = 0;
+	OSStatus err = AudioObjectGetPropertyDataSize(device_id, &aopa, 0, NULL, &size);
+	if (err != noErr)
+		return channels;
+
+	AudioBufferList* buffer_list = (AudioBufferList*)malloc(size);
+	err = AudioObjectGetPropertyData(device_id, &aopa, 0, NULL, &size, buffer_list);
+	if (err == noErr)
+	{
+		for(UInt32 buffer = 0; buffer < buffer_list->mNumberBuffers; ++buffer)
+			channels += buffer_list->mBuffers[buffer].mNumberChannels;
+	}
+	else
+	{
+		FormatError(macos_output_domain, "Unable to get total device output channels - id: 0x%04x. Error = %s", (uint)device_id, GetError(err).c_str());
+	}
+
+	free(buffer_list);
+
+	return channels;
+}
+
+UInt32
+CoreAudioDevice::GetNumChannelsOfStream(UInt32 stream_idx) const
+{
+	UInt32 channels = 0;
+
+	if (!device_id)
+		return channels;
+  
+	AudioObjectPropertyAddress  aopa;
+	aopa.mScope    = kAudioDevicePropertyScopeOutput;
+	aopa.mElement  = 0;
+	aopa.mSelector = kAudioDevicePropertyStreamConfiguration;
+
+	UInt32 size = 0;
+	OSStatus err = AudioObjectGetPropertyDataSize(device_id, &aopa, 0, NULL, &size);
+	if (err != noErr)
+		return channels;
+  
+	AudioBufferList* buffer_list = (AudioBufferList*)malloc(size);
+	err = AudioObjectGetPropertyData(device_id, &aopa, 0, NULL, &size, buffer_list);
+	if (err == noErr)
+	{
+		if (stream_idx < buffer_list->mNumberBuffers)
+			channels = buffer_list->mBuffers[stream_idx].mNumberChannels;
+	}
+	else
+	{
+	  FormatError(macos_output_domain, "Unable to get number of stream output channels - id: 0x%04x. Error = %s.",
```

Comment:
Throw exception?

Reply:
Yes, in general I changed a lot of the code to just throw instead of only printing errors